reasoner/Reasoner.java

```diff
@@ -77,56 +83,102 @@ ResolverRegistry resolverRegistry() {
-    }
+    public ResourceIterator<ConceptMap> execute(Disjunction disjunction, Set<Identifier.Variable.Name> filter,
+                                                @Nullable Long offset, @Nullable Long limit, Context.Query context) {
 
-    private boolean isInfer(Context.Query context) {
-        return context.options().infer() && !context.transactionType().isWrite() && logicMgr.rules().hasNext();
-    }
+        Disjunction satisfiableDisjunction = resolveTypesAndFilter(disjunction, filter);
+        if (satisfiableDisjunction.conjunctions().isEmpty()) return Iterators.empty();
+        if (isInfer(satisfiableDisjunction, context)) return resolve(satisfiableDisjunction, filter, offset, limit, context);
 
-    private boolean conjunctionContainsThings(Conjunction conjunction, Set<Identifier.Variable.Name> filter) {
-        return !filter.isEmpty() && iterate(filter).anyMatch(id -> conjunction.variable(id).isThing()) ||
-                iterate(conjunction.variables()).anyMatch(Variable::isThing);
-    }
-
-    public ResourceIterator<ConceptMap> execute(Disjunction disjunction, Set<Identifier.Variable.Name> filter,
-                                                Context.Query context) {
         ResourceIterator<ConceptMap> answers;
-        ResourceIterator<Conjunction> conjs = iterate(disjunction.conjunctions());
+        ResourceIterator<Conjunction> conjs = iterate(satisfiableDisjunction.conjunctions());
         if (!context.options().parallel()) answers = conjs.flatMap(conj -> iterator(conj, filter, context));
         else answers = produce(conjs.map(c -> producer(c, filter, context)).toList(), context.producer(), asyncPool1());
-        if (disjunction.conjunctions().size() > 1) answers = answers.distinct();
+        if (satisfiableDisjunction.conjunctions().size() > 1) answers = answers.distinct();
         return answers;
     }
 
-    private Producer<ConceptMap> producer(Conjunction conjunction) {
-        return producer(conjunction, set(), defaultContext);
+    private boolean isInfer(Disjunction disjunction, Context.Query context) {
+        if (!context.options().infer() || context.transactionType().isWrite() || !logicMgr.rules().hasNext()) {
+            return false;
+        }
+        for (Conjunction conj : disjunction.conjunctions()) {
+            for (Variable var : conj.variables()) {
+                if (var.resolvedTypes().isEmpty()) return true;
+                for (Label type : var.resolvedTypes()) {
+                    if (logicMgr.rulesConcluding(type).hasNext() || logicMgr.rulesConcludingHas(type).hasNext()) {
+                        return true;
+                    }
+                }
+            }
+            if (!conj.negations().isEmpty()) {
+                for (Negation negation : conj.negations()) {
+                    if (isInfer(negation.disjunction(), context)){
+                        return true;
+                    }
+                }
+            }
+        }
```

Comment:
we check if any rules are applicable to any part of the query, checking recursively